src/aggregation/metric/top_hits.rs

```diff
@@ -0,0 +1,350 @@
+        // `Ord` semantics of the heap elements.
+        hits.drain(..self.req.from.unwrap_or(0));
+        TopHitsMetricResult { hits }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct SegmentTopHitsCollector {
+    segment_id: SegmentOrdinal,
+    accessor_idx: usize,
+    inner_collector: TopHitsCollector,
+}
+
+impl SegmentTopHitsCollector {
+    // // TODO: confirm if this is required (idts)
+    // pub fn new(
+    //     segment_id: SegmentOrdinal,
+    //     accessor_idx: usize,
+    //     req: TopHitsAggregation,
+    // ) -> SegmentTopHitsCollector { SegmentTopHitsCollector { segment_id, accessor_idx,
+    //   inner_collector: TopHitsCollector { req, heap: BinaryHeap::new(), }, }
+    // }
+
+    pub fn from_req(
+        req: &TopHitsAggregation,
+        accessor_idx: usize,
+        segment_id: SegmentOrdinal,
+    ) -> Self {
+        Self {
+            inner_collector: TopHitsCollector {
+                req: req.clone(),
+                heap: BinaryHeap::with_capacity(req.size + req.from.unwrap_or(0)),
+            },
+            segment_id,
+            accessor_idx,
+        }
+    }
+}
+
+impl std::fmt::Debug for SegmentTopHitsCollector {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SegmentTopHitsCollector")
+            .field("segment_id", &self.segment_id)
+            .field("accessor_idx", &self.accessor_idx)
+            .field("inner_collector", &self.inner_collector)
+            .finish()
+    }
+}
+
+impl SegmentAggregationCollector for SegmentTopHitsCollector {
+    fn add_intermediate_aggregation_result(
+        self: Box<Self>,
+        agg_with_accessor: &crate::aggregation::agg_req_with_accessor::AggregationsWithAccessor,
+        results: &mut crate::aggregation::intermediate_agg_result::IntermediateAggregationResults,
+    ) -> crate::Result<()> {
+        let name = agg_with_accessor.aggs.keys[self.accessor_idx].to_string();
+        let intermediate_result = IntermediateMetricResult::TopHits(self.inner_collector);
+        results.push(
+            name,
+            IntermediateAggregationResult::Metric(intermediate_result),
+        )
+    }
+
+    fn collect(
+        &mut self,
+        doc_id: crate::DocId,
+        agg_with_accessor: &mut crate::aggregation::agg_req_with_accessor::AggregationsWithAccessor,
+    ) -> crate::Result<()> {
+        let features: Vec<ComparableDocFeature> = agg_with_accessor.aggs.values[self.accessor_idx]
+            .accessors
+            .iter()
+            .zip(self.inner_collector.req.sort.iter())
+            .map(|((c, _), KeyOrder(_, order))| {
+                let order = *order;
+                match c.values_for_doc(doc_id).next() {
+                    Some(value) => ComparableDocFeature { value, order },
+                    // TODO: confirm if this default 0-value is correct
+                    None => ComparableDocFeature { value: 0, order },
```

Comment:
This mixes 0 and non-existence which will be wrong in some cases

Reply:
How do you suggest handling missing fields then? Perhaps with a None and rank them below docs with extant values?

Reply:
Yes, including `None` and scoring it lower than anything with a value would be the right approach

Reply:
Done with [74654bf](https://github.com/quickwit-oss/tantivy/pull/2198/commits/74654bf48b7f61793aebaaacedcafc03f7c78c9d)